torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -513,6 +513,16 @@ class ShapePropagator {
         }
         return;
       }
+      case prim::unchecked_unwrap_optional: {
+	// we know we cannot have None as input, so we can always pass
+	// on the type.
+	if(auto ot = node->input()->type()->cast<OptionalType>()) {
+	  node->output()->setType(ot->getElementType());
+	} else {
+	  node->output()->setType(node->input()->type());
```

Comment:
Why would we have those calls on non-optional inputs?

Reply:
Also, when is that op used?

Reply:
These calls would be inserted before input types have been specialized. If we have an optional tensor as an input to the graph and we specialize the graph to a non-optional tensor, than these calls will no longer be needed

Reply:
When does an unchecked_unwrap_optional ever have a different type than ot->getElementType? This code looks like a no-op to me. Why is it ineeded?

Reply:
So as this PR is dead and has been reverted, I tried to add a better high-level overview this in #18407 .
The gist is that I've been trying to convert Optional[Tensor] into DimensionedTensorType or NoneType once we know what has been input.
May I suggest we move the discussion to #18407?